fix(compiler-cli): handle pre-release versions when checking version

crisbeto opened the pull request:
Currently the TS version checking function interprets a version like `1.2.3-rc.5` as `1.2.NaN` which would allow it to bypass the version checking altogether.

These changes add a little bit more logic to ensure that such versions are handled correctly. There's also an error if we don't manage to parse the version string.

Also it seemed like we never actually ran the version check unit tests, because they didn't have a test target.

alan-agius4 commented on packages/compiler-cli/src/diagnostics/typescript_version.ts:
```diff
@@ -12,7 +12,17 @@
  */
 export function toNumbers(value: string): number[] {
-  return value.split('.').map(Number);
+  // Drop any suffixes starting with `-` so that versions like `1.2.3-rc.5` are treated as `1.2.3`.
```
The compiler-cli depends on semver https://github.com/angular/angular/blob/e96f1653bd5f6987b51c5324284d92baf29bee52/packages/compiler-cli/package.json#L63 which can be used to simply this slightly. 

```ts
import { parse } from 'semver';

export function toNumbers(value: string): number[] {
  const version = parse(value);
  if (!version) {
    throw Error(`Unable to parse version string ${value}.`);
  }

  const {major, minor, patch} = version;
  return [major, minor, patch];
}
```

crisbeto replied:
If we're bringing in `semver`, why not just use `semver.satisfies(version, 'min - max')`?

JoostK replied:
`semver` is currently only used in code outside of g3 (i.e. the linker) and using it from g3-synced code may introduce a new dependency for g3.

crisbeto replied:
ok, this code is definitely checked into g3 so I'd rather avoid introducing a new dependency.

alan-agius4 replied:
Agreed

devversion commented on packages/compiler-cli/test/BUILD.bazel:
```diff
@@ -145,3 +145,25 @@ jasmine_node_test(
+    name = "typescript_support",
+    bootstrap = ["//tools/testing:node_es5"],
+    data = [
+        "//packages/core:npm_package",
```
it looks like you can remove this target as it is not used (and these NPM packages depend on a lot of things transitively)

crisbeto committed:
fix(compiler-cli): handle pre-release versions when checking version

Currently the TS version checking function interprets a version like `1.2.3-rc.5` as `1.2.NaN` which would allow it to bypass the version checking altogether.

These changes add a little bit more logic to ensure that such versions are handled correctly. There's also an error if we don't manage to parse the version string.

Also it seemed like we never actually ran the version check unit tests, because they didn't have a test target.

JoostK commented on packages/compiler-cli/src/diagnostics/typescript_version.ts:
```diff
@@ -12,7 +12,17 @@
+  // Drop any suffixes starting with `-` so that versions like `1.2.3-rc.5` are treated as `1.2.3`.
+  const suffixIndex = value.lastIndexOf('-');
+  return value.slice(0, suffixIndex === -1 ? value.length : suffixIndex).split('.').map(segment => {
+    const parsed = parseInt(segment);
```
To be on the safe side, should we pass an explicit radix 10 here?

JoostK commented on packages/compiler-cli/test/BUILD.bazel:
```diff
@@ -145,3 +145,22 @@ jasmine_node_test(
 )
+
+ts_library(
+    name = "typescript_support_lib",
```
Once VE stuff gets deleted we should be able to find a better home for this test in one of ngtsc's subpackages.

crisbeto committed:
fixup! fix(compiler-cli): handle pre-release versions when checking version

atscott commented:
This PR was merged into the repository by commit 247c18dee011f8a743b0b438b8ba2f017b113921.
